xarray/core/coordinates.py

```diff
@@ -366,7 +369,7 @@ def dtypes(self) -> Frozen[Hashable, np.dtype]:
     def _names(self) -> set[Hashable]:
         return set(self._data._coords)
 
-    def __getitem__(self, key: Hashable) -> DataArray:
+    def __getitem__(self, key: Hashable) -> T_DataArray:
         return self._data._getitem_coord(key)
```

Comment:
```python
xarray/core/coordinates.py:372: error: A function returning TypeVar should receive at least one argument containing the same Typevar  [type-var]
xarray/core/coordinates.py:372: note: Consider using the upper bound "DataArray" instead
```
I don't really understand this error, why should the method have at least one argument with the same TypeVar? The TypeVar is stored in self already.

Reply:
That's what I meant with making coordinates a generic class.

Reply:
Sorry, I'm not following. Feel free to push to this PR or link me to an example you think is similar.

Reply:
done :)

Reply:
Ok, nvmd. This does not seem to work in python 3.8....
Anyone has an idea how to solve this? haha.

Reply:
I've been googleing without much luck. :(

One super ugly way is to just do a gigantic `if sys.version_info >= (3, 9)` check switching between the new and old version.

Reply:
Yes, it is extremely difficult to Google python version dependent stuff...

Reply:
Wow that is an interesting hack, haha

Reply:
Yeah, a little surprised it worked. :D

Taken from:
https://github.com/python/cpython/blob/b0e1f9c241cd8f8c864d51059217f997d3b792bf/Lib/_collections_abc.py#L428

Reply:
Does it work though? What does a `reveal_type(da.coords)` give?
Or a `reveal_type(da.coords["x"])`?

Reply:
I get this on main:
```python
da = xr.DataArray()
reveal_type(da)  # note: Revealed type is "Any"
```

Do you as well?

Reply:
No, I get `note: Revealed type is "xarray.core.dataarray.DataArray"` as expected.

Reply:
Hmm, I use:
```
mypy                      0.982           py310h8d17308_0    conda-forge
mypy_extensions           0.4.3           py310h5588dad_5    conda-forge
...
python                    3.10.6          h9a09f29_0_cpython    conda-forge
```